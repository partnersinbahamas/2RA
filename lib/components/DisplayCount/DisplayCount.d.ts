import { default as TArticleProps } from '../utils/types/props/article';
import { TStile } from '../utils/types/types';

type TProps = TArticleProps & {
    text: string;
    x: number | string;
    y: number | string;
    fixTo?: string;
    stile?: TStile;
    className?: any;
};
declare const DisplayCount: React.FC<TProps>;
export default DisplayCount;
