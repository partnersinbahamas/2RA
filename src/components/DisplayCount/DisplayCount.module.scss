@import '../utils/styles/styles.scss';

.article {
  &-default {
    @extend %fontTuraRegular;

    &-variables {
      padding: 0.1rem;
      background-color: $cool_gray_200;
      border-bottom: 2px solid $cool_gray_300;
    }
  }

  &-primary {
    @extend %fontTuraRegular;

    &-variables {
      padding: 0.1rem;
      background-color: $blue_300;
      color: $white;
      border-bottom: 2px solid $blue_400;
    }
  }
}
