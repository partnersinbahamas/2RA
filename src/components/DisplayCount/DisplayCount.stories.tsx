import { Meta, StoryObj } from '@storybook/react';
import DisplayCount from './DisplayCount';
import { story } from '../utils/variables/story';
import { Description, DocsPage, Source } from '@storybook/blocks';

const DISPLAY_COUNT_MODULES_SCSS_STYLES = `
.displayCount {
  .displayCount-variables {};
  .displayCount-x {};
  .displayCount-y {};
}
`;

const DISPLAY_COUNT_SCSS_STYLES = `
.displayCount {  
  &-variables {};
  &-x {};
  &-y {};
}
`;

const meta = {
  title: 'Displays/DisplayCount',
  component: DisplayCount,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
    docs: {
      page: () => {
        return (
          <>
            <Description>
              {`/* Please to style your custom DisplayCount component use the mockup below. */`}
            </Description>

            <Description markdown={`## Guide styles | .module.scss`} />
            <Source language="css" code={DISPLAY_COUNT_MODULES_SCSS_STYLES} />

            <Description markdown={`## Guide styles | .scss`} />
            <Source language="css" code={DISPLAY_COUNT_SCSS_STYLES} />
            <DocsPage />
          </>
        );
      },
    },
  },
  args: {
    text: 'Display X of Y',
    x: 10,
    y: 0,
  },
  argTypes: {
    stile: story.stile,
    text: { control: 'text' },
    x: { control: 'text' } || { control: 'number' },
    y: { control: 'text' } || { control: 'number' },
    fixTo: { control: 'text' },
    className: story.className,
  },
} satisfies Meta<typeof DisplayCount>;
export default meta;

type TStory = StoryObj<typeof meta>;

export const Default: TStory = { args: {} };
export const Primary: TStory = { args: { stile: 'primary' } };
export const mute: TStory = { args: { stile: 'mute' } };
