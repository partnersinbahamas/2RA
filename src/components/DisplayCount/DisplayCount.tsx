import PropsTypes from 'prop-types';
import classNames from 'classnames';

import { useHD } from '../../providers/HDProvider';
import useModuleExtention from '../../hooks/useModuleExtention';

import TArticleProps from '../utils/types/props/article';
import { numberFixTo } from '../utils/functions';
import defaultProps from '../utils/variables/defaultProps';
import { TStile } from '../utils/types/types';
import styles from './DisplayCount.module.scss';

type TProps = TArticleProps & {
  text: string;
  x: number | string;
  y: number | string;
  fixTo?: string;
  stile?: TStile;
  className?: any;
};

const DisplayCount: React.FC<TProps> = ({
  text,
  x,
  y,
  fixTo,
  stile = defaultProps.stile,
  className,
}) => {
  const { stylesExtention } = useHD();
  const modulesExtension = useModuleExtention(
    stylesExtention as TStylesExtension,
  ).moduleExtentionState;

  const modifiedX = fixTo?.length ? numberFixTo(x, fixTo) : x;
  const modifiedY = fixTo?.length ? numberFixTo(y, fixTo) : y;

  const stiles = {
    article: styles[`article-${stile}`],
    variables: styles[`article-${stile}-variables`],
    x: styles[`article-${stile}-x`],
    y: styles[`article-${stile}-y`],
  };

  const classes = {
    displayCount:
      className && modulesExtension ? className.displayCount : className,
    variables:
      className && modulesExtension
        ? className['displayCount-variables']
        : `${className}-variables`,
    x:
      className && modulesExtension
        ? className['displayCount-x']
        : `${className}-x`,
    y:
      className && modulesExtension
        ? className['displayCount-y']
        : `${className}-y`,
  };

  const modifiedText = text
    .replace(
      /\bX+\b/g,
      `<strong class="${stiles.x} ${stiles.variables} ${classes.variables} ${classes.x}" data-testid="displayCount-x">${modifiedX}</strong>`,
    )
    .replace(
      /\bY+\b/g,
      `<strong class="${stiles.y} ${stiles.variables} ${classes.variables} ${classes.y}" data-testid="displayCount-y">${modifiedY}</strong>`,
    );

  const messageHTML = { __html: modifiedText };

  return (
    <article
      className={classNames(stiles.article, classes.displayCount)}
      dangerouslySetInnerHTML={messageHTML}
    />
  );
};

DisplayCount.propTypes = {
  text: PropsTypes.string.isRequired,
  x: PropsTypes.oneOfType([PropsTypes.number, PropsTypes.string]).isRequired,
  y: PropsTypes.oneOfType([PropsTypes.number, PropsTypes.string]).isRequired,
  fixTo: PropsTypes.string,
  className: PropsTypes.any,

  // any cause null not exist in PropsTypes
  stile: PropsTypes.oneOfType([PropsTypes.string, PropsTypes.any]),
};

export default DisplayCount;
