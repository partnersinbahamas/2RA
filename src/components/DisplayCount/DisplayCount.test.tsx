/* eslint-disable no-unsafe-optional-chaining */
import { render, screen } from '@testing-library/react';
import DisplayCount from './DisplayCount';

const displayCountFixToFormats = [
  'X',
  'XX',
  'XXX',
  'XXXX',
  'XXXXX',
  'XXXXXX',
  'XXXXXXX',
  'XXXXXXXX',
  'XXXXXXXXX',
  'XXXXXXXXXX',
  'XXXXXXXXXXX',
  'XXXXXXXXXXXX',
  'XXXXXXXXXXXXX',
  'XXXXXXXXXXXXXX',
  'XXXXXXXXXXXXXXX',
  'XXXXXXXXXXXXXXXX',
];

describe('DisplayCount', () => {
  const defaultProps = {
    text: 'Display X / Y',
    x: 0.7,
    y: 1,
  };

  const { x, y } = defaultProps;

  test.each(['Display X / Y', 'Display XXX / YYY'])(
    'renders correctly',
    text => {
      const { container } = render(
        <DisplayCount {...defaultProps} text={text} />,
      );

      const articleElement = screen.getByRole('article');
      expect(articleElement).toBeInTheDocument();
      expect(articleElement.textContent).toBe(`Display ${x} / ${y}`);

      expect(container).toMatchSnapshot();
    },
  );

  test.each(displayCountFixToFormats)('renders with fixTo', fixTo => {
    const { container } = render(
      <DisplayCount {...defaultProps} fixTo={fixTo} />,
    );

    const numXtxt = screen.getByTestId('displayCount-x').textContent;
    const numYtxt = screen.getByTestId('displayCount-y').textContent;

    if (!numXtxt || !numYtxt) return;

    const [_X, decimalPartX] = numXtxt?.split('.');
    const [_Y, decimalPartY] = numYtxt?.split('.');

    expect(decimalPartX.length).toBe(fixTo.length);
    expect(decimalPartY.length).toBe(fixTo.length);

    expect(container).toMatchSnapshot();
  });

  test('renders with x and y in lower case', () => {
    render(<DisplayCount {...defaultProps} text="Display x / y" />);
    expect(screen.getByRole('article').textContent).toBe(`Display x / y`);
  });
});
