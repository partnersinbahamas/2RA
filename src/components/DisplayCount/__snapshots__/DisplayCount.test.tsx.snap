// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayCount renders correctly 1`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders correctly 2`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 1`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.0
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 2`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.70
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.00
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 3`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.700
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 4`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.0000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 5`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.70000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.00000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 6`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.700000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 7`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.0000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 8`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.70000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.00000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 9`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.700000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 10`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.0000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 11`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.70000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.00000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 12`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.700000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.000000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 13`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7000000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.0000000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 14`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.70000000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.00000000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 15`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.700000000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.000000000000000
    </strong>
  </article>
</div>
`;

exports[`DisplayCount renders with fixTo 16`] = `
<div>
  <article
    class="article-default"
  >
    Display 
    <strong
      class="article-default-x article-default-variables undefined-variables undefined-x"
      data-testid="displayCount-x"
    >
      0.7000000000000000
    </strong>
     / 
    <strong
      class="article-default-y article-default-variables undefined-variables undefined-y"
      data-testid="displayCount-y"
    >
      1.0000000000000000
    </strong>
  </article>
</div>
`;
